### PR TITLE
Null handling in SELECT, ORDER BY, DISTINCT, HAVING, GROUP BY, and few agg. functions and few data types

### PR DESCRIPTION
**Pull request by nizarhejazi**
https://github.com/apache/pinot/issues/8697

This is a draft PR that demonstrates NULL value support in:

1. SELECT, ORDER BY, DISTINCT, HAVING, GROUP BY.
2. Aggregation functions: COUNT, MAX.
3. Non-dictionary columns.
4. Basic transforms.
5. Few data types: Boolean, Int, and BigDecimal.

Note: updated `BigDecimalQueriesTest` to work with null values. Introduced `BooleanNullEnabledNoDictQueriesTest`. Total: 21 tests with nulls.

Once this is reviewed, if reviewers agree with my approach, I will send a PR soon and follow up with few more PRs to get NULL support done for:

- All SELECT statements
- All data types
- Dictionary/non-dictionary based columns
- All transforms

`feature`, `release-notes`,

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java [hunk 1]
```diff
@@ -79,14 +79,30 @@ public TableResizer(DataSchema dataSchema, QueryContext queryContext) {
     _numOrderByExpressions = orderByExpressions.size();
     _orderByValueExtractors = new OrderByValueExtractor[_numOrderByExpressions];
     Comparator[] comparators = new Comparator[_numOrderByExpressions];
+    int[] comparisonSign = new int[_numGroupByExpressions];
```
Comment (nizarhejazi) [inline pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java:82]: Found that in Presto: "The default null ordering is NULLS LAST, **regardless of the ordering direction**". Will update the behviour in Pinot to match it.

Comment (nizarhejazi) [inline pinot-core/src/main/java/org/apache/pinot/core/data/table/TableResizer.java:82]: Found that in Presto: "The default null ordering is NULLS LAST, **regardless of the ordering direction**". Will update the behviour in Pinot to match it.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java [hunk 2]
```diff
@@ -173,6 +188,25 @@ public byte[] toBytes()
       dataOutputStream.writeInt(bytes.length);
       dataOutputStream.write(bytes);
     }
+
+    // Write colum field specs.
+    // todo: consider serializing only defaultNullValues since this is the only required info.
+    int fieldSpecsSize = _columnFieldSpecs == null ? 0 : _columnFieldSpecs.length;
+    dataOutputStream.writeInt(fieldSpecsSize);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:195]: (MAJOR) This is backward incompatible change. The new broker won't understand the data schema returned from the old servers

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java [hunk 3]
```diff
@@ -46,6 +49,7 @@
 public class DataSchema {
   private final String[] _columnNames;
   private final ColumnDataType[] _columnDataTypes;
+  private final FieldSpec[] _columnFieldSpecs;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:52]: Trying to understand why we need `FieldSpec` in the data schema. Seems the intention is to use it to calculate the default null value.
It should not be stored here for the following reasons:
1. We don't want to carry the field spec in the transport object
2. The column could be an expression (transform function/aggregation function etc), and there is no existing `FieldSpec` for them
3. We should not need to calculate default null value on the fly. Filling any value should be fine because we should skip processing them

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java [hunk 4]
```diff
@@ -76,6 +77,8 @@ byte[] toBytes()
 
   String[] getStringArray(int rowId, int colId);
 
+  MutableRoaringBitmap getColumnNullBitmap(int colId);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java:80]: (minor) Suggest renaming to `getNullRowIds()`

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableBuilder.java [hunk 5]
```diff
@@ -102,11 +110,17 @@ public DataTableBuilder(DataSchema dataSchema) {
   }
 
   public static DataTable getEmptyDataTable() {
-    return _version == VERSION_2 ? new DataTableImplV2() : new DataTableImplV3();
+    if (_version == VERSION_2) {
+      return new DataTableImplV2();
+    }
+    if (_version == VERSION_4) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableBuilder.java:116]: (minor) Suggest checking `VERSION_3` for readability, same for `build()`

File: pinot-core/src/main/java/org/apache/pinot/core/query/distinct/raw/BaseRawIntSingleColumnDistinctExecutor.java [hunk 6]
```diff
@@ -29,35 +28,34 @@
...
  * Base implementation of {@link DistinctExecutor} for single raw INT column.
  */
 abstract class BaseRawIntSingleColumnDistinctExecutor implements DistinctExecutor {
   final ExpressionContext _expression;
-  final DataType _dataType;
+  final FieldSpec _fieldSpec;
   final int _limit;
 
-  final IntSet _valueSet;
+  final ObjectSet<Integer> _valueSet;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/distinct/raw/BaseRawIntSingleColumnDistinctExecutor.java:42]: This can cause performance regression. We should avoid adding overhead to existing use cases. If we want to return `null` as a separate distinct value, we can add a boolean flag `_hasNull`

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/DistinctDataTableReducer.java [hunk 7]
```diff
@@ -99,11 +99,11 @@ private ResultTable reduceToResultTable(DistinctTable distinctTable) {
     Iterator<Record> iterator = distinctTable.getFinalResult();
     while (iterator.hasNext()) {
       Object[] values = iterator.next().getValues();
-      Object[] row = new Object[numColumns];
+      Object[] rowValues = new Object[numColumns];
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/DistinctDataTableReducer.java:102]: (minor) Seems unrelated

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java [hunk 8]
```diff
@@ -136,23 +137,23 @@ private void reduceToResultTable(BrokerResponseNative brokerResponseNative, Data
     if (havingFilter != null) {
       HavingFilterHandler havingFilterHandler = new HavingFilterHandler(havingFilter, postAggregationHandler);
       while (rows.size() < limit && sortedIterator.hasNext()) {
-        Object[] row = sortedIterator.next().getValues();
-        extractFinalAggregationResults(row);
+        Object[] values = sortedIterator.next().getValues();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java:140]: (minor) Suggest reverting it. I feel `row` is more specific, `values` is too generic

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/BigDecimalColumnPreIndexStatsCollector.java [hunk 9]
```diff
@@ -112,8 +113,13 @@ public int getCardinality() {
   }
 
   @Override
-  public boolean hasNull() {
-    return false;
+  public boolean hasNulls() {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/BigDecimalColumnPreIndexStatsCollector.java:116]: This API is not really used. We should rely on whether null value vector exist and non-empty to identify whether a column contains null values

**Review by Jackie-Jiang — COMMENTED** [review]
The approach is in general good. We should avoid adding overhead (minimal overhead of a flag check is good, but changing primitive value to boxed value can cause performance regression) to existing use cases if null handling is not enabled.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/BigDecimalColumnPreIndexStatsCollector.java (hunk 9 above)
Comment (nizarhejazi) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/BigDecimalColumnPreIndexStatsCollector.java:116]: Attempted to serialized whether dictionary has null value but found that dictionary serialization format cannot be changed easily.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 2 above)
Comment (nizarhejazi) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:195]: Will take care of ensuring backward compatibility in final PR .. this is just a draft PR to get the discussion going.
